### PR TITLE
fix(mixpanel): batch event ordering

### PR DESCRIPTION
**Pull request by Gauravudia**
## Description of the change

> Resolves INT-390
Fix mixpanel  batch even ordering and move utility functions to utils.js

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/mp/util.js [hunk 1]
```diff
@@ -166,6 +170,102 @@ const removeDuplicateMetadata = (mergedBatches) => {
...
+  const importEvents = [];
+  const batchErrorRespList = [];
+  events.forEach((result) => {
+    if (result.message) {
+      const { destination, metadata } = result;
+      let { message } = result;
+      message = CommonUtils.toArray(message);
+      message.forEach((msg) => {
+        // eslint-disable-next-line default-case
+        switch (true) {
```
Comment (chandumlg) [inline src/v0/destinations/mp/util.js:223]: Isn't there a better way to do this?

File: src/v0/destinations/mp/transform.js [hunk 2]
```diff
@@ -470,28 +382,56 @@ const processRouterDest = async (inputs, reqMetadata) => {
...
+        ...importRespList,
+      ];
+
+      return [...batchSuccessRespList, ...batchErrorRespList];
+    }),
+  );
+
+  // Flatten the response array containing batched events from multiple groups
+  const allBatchedEvents = _.flatMap(response);
+  return combineBatchRequestsWithSameJobIds(allBatchedEvents);
```
Comment (chandumlg) [inline src/v0/destinations/mp/transform.js:434]: Why should we do this? Why would jobIds repeat?

File: src/v0/destinations/mp/util.js [hunk 3]
```diff
@@ -166,6 +170,102 @@ const removeDuplicateMetadata = (mergedBatches) => {
...
+    batchErrorRespList,
+  };
+};
+
+const generateBatchedPayloadForArray = (events) => {
+  const { batchedRequest } = defaultBatchRequestConfig();
+  const batchResponseList = events.flatMap((event) => JSON.parse(event.body.JSON_ARRAY.batch));
+  batchedRequest.body.JSON_ARRAY = { batch: JSON.stringify(batchResponseList) };
+  batchedRequest.endpoint = events[0].endpoint;
+  batchedRequest.headers = events[0].headers;
```
Comment (chandumlg) [inline src/v0/destinations/mp/util.js:256]: Question: Does this events array contain events of different destinations? If yes, is  that okay?

File: src/v0/destinations/mp/transform.js (hunk 2 above)
Comment (Gauravudia) [reply src/v0/destinations/mp/transform.js:434]: In identify, group and track calls we are doing the multiplexing of events. So, two events can have the same JobIds.

File: src/v0/destinations/mp/util.js (hunk 3 above)
Comment (Gauravudia) [reply src/v0/destinations/mp/util.js:256]: Events array will contain the events of same destination.

File: src/v0/destinations/mp/util.js (hunk 1 above)
Comment (Gauravudia) [reply src/v0/destinations/mp/util.js]: refactor the code and replaced the switch case and used .find()

**Review by shrouti1507 — APPROVED** [review]

**Pull request merged**